tsl/src/continuous_aggs/insert.c

```diff
@@ -324,9 +335,11 @@ continuous_agg_xact_invalidation_callback(XactEvent event, void *arg)
 	switch (event)
 	{
 		case XACT_EVENT_PRE_COMMIT:
+		case XACT_EVENT_PARALLEL_PRE_COMMIT:
 			cache_inval_htab_write();
-			cache_inval_cleanup();
 			break;
+		case XACT_EVENT_COMMIT:
+		case XACT_EVENT_PARALLEL_COMMIT:
 		case XACT_EVENT_ABORT:
 		case XACT_EVENT_PARALLEL_ABORT:
 			cache_inval_cleanup();
```

Comment:
The events `XACT_PREPARE` and `XACT_PRE_PREPARE` are missing. Could you explain why they are not needed?

Reply:
Yes, we need to fix that.

Reply:
Just to summarize our discussion:

During execution of an insert into a hypertable, either as part of a transaction or as a single statement, the hypertable invalidations are collected into a hash table `continuous_aggs_cache_inval_htab` through the trigger function for the hypertable `continuous_agg_trigfn`. This hash table is ephemeral and will not survive a crash.

When executing 2PC after inserting into a hypertable that has a continuous aggregate, the steps are as follows:
1. `XACT_EVENT_PRE_PREPARE`: Nothing specific is done.
2. `XACT_EVENT_PREPARE`: Nothing specific is done.
3. `XACT_EVENT_PRE_COMMIT`: In `continuous_agg_xact_invalidation_callback` there is a switch that will write the hash table to the snapshot.
4. `XACT_EVENT_COMMIT`: Nothing specific is done.

After step 2, the data node will report that all is durable and the access node can commit the transaction. However, this is not true. If there is a crash in one of the data nodes, the hash table will be empty after recovery, which means that the invalidations that were reported durable in step 2 cannot be written to the snapshot in step 3 and will be lost. The solution is to write the hash table to the snapshot (hence also to the WAL) in step 1 instead of step 3.